xds/src/main/java/io/grpc/xds/LocalityStore.java

```diff
@@ -0,0 +1,401 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.CONNECTING;
+import static io.grpc.ConnectivityState.IDLE;
+import static io.grpc.ConnectivityState.READY;
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.grpc.Attributes;
+import io.grpc.ChannelLogger.ChannelLogLevel;
+import io.grpc.ConnectivityState;
+import io.grpc.ConnectivityStateInfo;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancer.Helper;
+import io.grpc.LoadBalancer.PickResult;
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.LoadBalancer.ResolvedAddresses;
+import io.grpc.LoadBalancer.Subchannel;
+import io.grpc.LoadBalancer.SubchannelPicker;
+import io.grpc.LoadBalancerProvider;
+import io.grpc.Status;
+import io.grpc.util.ForwardingLoadBalancerHelper;
+import io.grpc.xds.InterLocalityPicker.WeightedChildPicker;
+import io.grpc.xds.XdsLbState.Locality;
+import io.grpc.xds.XdsLbState.LocalityInfo;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Manages EAG and locality info for a collection of subchannels, not including subchannels
+ * created by the fallback balancer.
+ */
+// Must be accessed/run in SynchronizedContext.
+interface LocalityStore {
+
+  boolean hasReadyBackends();
+
+  boolean hasNonDropBackends();
+
+  void shutdown();
+
+  void updateLocalityStore(Map<Locality, LocalityInfo> localityInfoMap);
+
+  void updateLoadBalancerProvider(LoadBalancerProvider loadBalancerProvider);
+
+  void handleSubchannelState(Subchannel subchannel, ConnectivityStateInfo newState);
```

Comment:
The method `handleSubchannelState` is here for now is because we are using the old deprecated API, not `SubchannelStateListener` yet. Removed `STATE_INFO` and `ChildHelper.subchannels`.